README.md
# cs486-project
## First Time Setup
In the root directory, run
```
bash setup.sh
```
which unzips the 1.6m training data and runs preprocessing as well as fetches covid-19 validation data from twitter

After the script finishes, you should get the following files in the `data` folder:
good_tweets.csv - positively labelled training data
bad_tweets.csv - negatively labelled training data
good_validation.csv - positively labelled Covid-19 validation data
bad_validation.csv - negatively labelled Covid-19 validation data


milestone2/changes.txt
We only changed the algorithms section to add more technical details for our chosen algorithm.


NN-algo/cnn.py
#####
# SOURCE: DENNY BRITZ - http://www.wildml.com/2015/12/implementing-a-cnn-for-text-classification-in-tensorflow/
#####

import tensorflow as tf
import numpy as np


class CNN(object):
    """
    A CNN for text classification.
    Uses an embedding layer, followed by a convolutional, max-pooling and softmax layer.
    """
    def __init__(
      self, sequence_length, num_classes, vocab_size,
      embedding_size, filter_sizes, num_filters, l2_reg_lambda=0.0):

        # Placeholders for input, output and dropout
        self.input_x = tf.placeholder(tf.int32, [None, sequence_length], name="input_x")
        self.input_y = tf.placeholder(tf.float32, [None, num_classes], name="input_y")
        self.dropout_keep_prob = tf.placeholder(tf.float32, name="dropout_keep_prob")

        # Keeping track of l2 regularization loss (optional)
        l2_loss = tf.constant(0.0)

        # Embedding layer
        with tf.device('/cpu:0'), tf.name_scope("embedding"):
            self.W = tf.Variable(
                tf.random_uniform([vocab_size, embedding_size], -1.0, 1.0),
                name="W")
            self.embedded_chars = tf.nn.embedding_lookup(self.W, self.input_x)
            self.embedded_chars_expanded = tf.expand_dims(self.embedded_chars, -1)

        # Create a convolution + maxpool layer for each filter size
        pooled_outputs = []
        for i, filter_size in enumerate(filter_sizes):
            with tf.name_scope("conv-maxpool-%s" % filter_size):
                # Convolution Layer
                filter_shape = [filter_size, embedding_size, 1, num_filters]
                W = tf.Variable(tf.truncated_normal(filter_shape, stddev=0.1), name="W")
                b = tf.Variable(tf.constant(0.1, shape=[num_filters]), name="b")
                conv = tf.nn.conv2d(
                    self.embedded_chars_expanded,
                    W,
                    strides=[1, 1, 1, 1],
                    padding="VALID",
                    name="conv")
                # Apply nonlinearity
                h = tf.nn.relu(tf.nn.bias_add(conv, b), name="relu")
                # Maxpooling over the outputs
                pooled = tf.nn.max_pool(
                    h,
                    ksize=[1, sequence_length - filter_size + 1, 1, 1],
                    strides=[1, 1, 1, 1],
                    padding='VALID',
                    name="pool")
                pooled_outputs.append(pooled)

        # Combine all the pooled features
        num_filters_total = num_filters * len(filter_sizes)
        self.h_pool = tf.concat(pooled_outputs, 3)
        self.h_pool_flat = tf.reshape(self.h_pool, [-1, num_filters_total])

        # Add dropout
        with tf.name_scope("dropout"):
            self.h_drop = tf.nn.dropout(self.h_pool_flat, self.dropout_keep_prob)

        # Final (unnormalized) scores and predictions
        with tf.name_scope("output"):
            W = tf.get_variable(
                "W",
                shape=[num_filters_total, num_classes],
                initializer=tf.contrib.layers.xavier_initializer())
            b = tf.Variable(tf.constant(0.1, shape=[num_classes]), name="b")
            l2_loss += tf.nn.l2_loss(W)
            l2_loss += tf.nn.l2_loss(b)
            self.scores = tf.nn.xw_plus_b(self.h_drop, W, b, name="scores")
            self.predictions = tf.argmax(self.scores, 1, name="predictions")

        # CalculateMean cross-entropy loss
        with tf.name_scope("loss"):
            losses = tf.nn.softmax_cross_entropy_with_logits(logits=self.scores, labels=self.input_y)
            self.loss = tf.reduce_mean(losses) + l2_reg_lambda * l2_loss

        # Accuracy
        with tf.name_scope("accuracy"):
            correct_predictions = tf.equal(self.predictions, tf.argmax(self.input_y, 1))
            self.accuracy = tf.reduce_mean(tf.cast(correct_predictions, "float"), name="accuracy")
        print ("LOADED CNN!")


NN-algo/batchgen.py
import csv
import re
import random
import numpy as np

from IPython import embed

#Separates a file with mixed positive and negative examples into two.
def separate_dataset(filename):
    good_out = open("good_"+filename,"w+");
    bad_out  = open("bad_"+filename,"w+");

    seen = 1;
    with open(filename,'r') as f:
        reader = csv.reader(f)
        reader.next()

        for line in reader:
            seen +=1
            sentiment = line[1]
            sentence = line[3]

            if (sentiment == "0"):
                bad_out.write(sentence+"\n")
            else:
                good_out.write(sentence+"\n")

            if (seen%10000==0):
                print (seen);

    good_out.close();
    bad_out.close();



#Load Dataset
def get_dataset(goodfile,badfile,limit,randomize=True):
    good_x = list(open(goodfile,"r", encoding = "ISO-8859-1").readlines())
    good_x = [s.strip() for s in good_x]
    
    bad_x  = list(open(badfile,"r", encoding = "ISO-8859-1").readlines())
    bad_x  = [s.strip() for s in bad_x]

    if (randomize):
        random.shuffle(bad_x)
        random.shuffle(good_x)

    good_x = good_x[:limit]
    bad_x = bad_x[:limit]

    x = good_x + bad_x
    x = [clean_str(s) for s in x]


    positive_labels = [[0, 1] for _ in good_x]
    negative_labels = [[1, 0] for _ in bad_x]
    y = np.concatenate([positive_labels, negative_labels], 0)
    return [x,y]




#Clean Dataset
def clean_str(string):


    #EMOJIS
    string = re.sub(r":\)","emojihappy1",string)
    string = re.sub(r":P","emojihappy2",string)
    string = re.sub(r":p","emojihappy3",string)
    string = re.sub(r":>","emojihappy4",string)
    string = re.sub(r":3","emojihappy5",string)
    string = re.sub(r":D","emojihappy6",string)
    string = re.sub(r" XD ","emojihappy7",string)
    string = re.sub(r" <3 ","emojihappy8",string)

    string = re.sub(r":\(","emojisad9",string)
    string = re.sub(r":<","emojisad10",string)
    string = re.sub(r":<","emojisad11",string)
    string = re.sub(r">:\(","emojisad12",string)

    #MENTIONS "(@)\w+"
    string = re.sub(r"(@)\w+","mentiontoken",string)
    
    #WEBSITES
    string = re.sub(r"http(s)*:(\S)*","linktoken",string)

    #STRANGE UNICODE \x...
    string = re.sub(r"\\x(\S)*","",string)

    #General Cleanup and Symbols
    string = re.sub(r"[^A-Za-z0-9(),!?\'\`]", " ", string)
    string = re.sub(r"\'s", " \'s", string)
    string = re.sub(r"\'ve", " \'ve", string)
    string = re.sub(r"n\'t", " n\'t", string)
    string = re.sub(r"\'re", " \'re", string)
    string = re.sub(r"\'d", " \'d", string)
    string = re.sub(r"\'ll", " \'ll", string)
    string = re.sub(r",", " , ", string)
    string = re.sub(r"!", " ! ", string)
    string = re.sub(r"\(", " \( ", string)
    string = re.sub(r"\)", " \) ", string)
    string = re.sub(r"\?", " \? ", string)
    string = re.sub(r"\s{2,}", " ", string)

    return string.strip().lower()



#Generate random batches
#Source: https://github.com/dennybritz/cnn-text-classification-tf/blob/master/data_helpers.py
def gen_batch(data, batch_size, num_epochs, shuffle=True):
    """
    Generates a batch iterator for a dataset.
    """
    data = np.array(data)
    data_size = len(data)
    num_batches_per_epoch = int((len(data)-1)/batch_size) + 1
    for epoch in range(num_epochs):
        # Shuffle the data at each epoch
        if shuffle:
            shuffle_indices = np.random.permutation(np.arange(data_size))
            shuffled_data = data[shuffle_indices]
        else:
            shuffled_data = data
        for batch_num in range(num_batches_per_epoch):
            start_index = batch_num * batch_size
            end_index = min((batch_num + 1) * batch_size, data_size)
            yield shuffled_data[start_index:end_index]

if __name__ == "__main__":
    separate_dataset("small.txt");


#42
#642


data_fetch/fetch.py
#!/usr/bin/python3

import sys
import os
import GetOldTweets3 as got
import pandas as pd
import csv
import re

CORONA_WORDS = [
    'social distancing', 'quarantine', 'self-isolation', 'covid19',
    'coronavirus'
]
SYMBOLS = ['@', '#', '"', ':', '-', ')', '(', ',', '.']
POSITIVE_EMOTICONS = [':)', ':-)']
NEGATIVE_EMOTICONS = [':(', ':-(']
SINCE = '2020-02-15'
UNTIL = '2020-05-01'
OUTPUT_FILE = 'data_fetch/validation_set_raw.csv'
GOOD_FILE = 'data/good_validation.csv'
BAD_FILE = 'data/bad_validation.csv'
LOCATION = 'US'
MAX = 10000


def build_query(main_word):
    q = main_word + ' (' + ' OR '.join(POSITIVE_EMOTICONS +
                                       NEGATIVE_EMOTICONS) + ')'
    return q


def run_download():
    all_tweets = pd.DataFrame(columns=['text'])
    for main_query_word in CORONA_WORDS:
        tweetCriteria = got.manager.TweetCriteria().setQuerySearch(
            build_query(main_query_word)).setSince(SINCE).setUntil(
                UNTIL).setMaxTweets(MAX).setNear(LOCATION).setWithin('3881mi')
        tweets = pd.DataFrame(map(
            lambda tweet: [tweet.id, tweet.text, tweet.date],
            got.manager.TweetManager.getTweets(tweetCriteria)),
                              columns=['tweet_id', 'text', 'date'])
        tweets = tweets.drop('tweet_id', axis=1).drop('date', axis=1)
        all_tweets = all_tweets.append(tweets, ignore_index=True)
    all_tweets.to_csv(OUTPUT_FILE, index=False)


def isEnglish(s):
    try:
        s.encode(encoding='utf-8').decode('ascii')
    except UnicodeDecodeError:
        return False
    else:
        return True


def run_preprocess():
    # 1. load csv
    data = pd.read_csv(OUTPUT_FILE)
    data['label'] = 0
    # 2. add label
    for i, row in data.iterrows():
        if any(ele in row['text'] for ele in POSITIVE_EMOTICONS):
            data.at[i, 'label'] = 1
    # 3. remove URLs
    for i, row in data.iterrows():
        data.at[i, 'text'] = re.sub(r"http\S+", "", data.at[i, 'text'])
    # 4. remove emoticons and symbols
    for i, row in data.iterrows():
        for e in (POSITIVE_EMOTICONS + NEGATIVE_EMOTICONS + SYMBOLS):
            data.at[i, 'text'] = data.at[i, 'text'].replace(e, '')
    # # 5. remove non-english tweets
    # for i, row in data.iterrows():
    #     if not isEnglish(row['text']):
    #         data.drop(i, inplace=True)

    # 6. split data into good_validation.csv bad_validation.csv
    good = data[data['label'] == 1].drop('label', axis=1)
    bad = data[data['label'] == 0].drop('label', axis=1)
    # 7. output stats
    print(f'outputting {len(good)} good tweets and {len(bad)} bad tweets')
    if not os.path.exists('../data'):
        os.makedirs('../data')
    good.to_csv(GOOD_FILE, index=False, header=False, quoting=csv.QUOTE_NONE)
    bad.to_csv(BAD_FILE, index=False, header=False, quoting=csv.QUOTE_NONE)


def main(argv):
    print('downloading data')
    run_download()
    print('running preprocessing')
    run_preprocess()


if __name__ == "__main__":
    main(sys.argv[1:])


NN-algo/graph_res.py
import matplotlib.pyplot as plt
import os


def plot_results(training_steps, training_losses, training_accuracy,
                 evaluation_steps, evaluation_losses, evaluation_accuracy,
                 evaluation_losses_noncovid, evaluation_accuracy_noncovid,
                 xmax, ymax, xmax2, ymax2):

    fig, graphs = plt.subplots(3, 2)
    fig.canvas.set_window_title('Results')
    plt.subplots_adjust(wspace=1.0, hspace=1.0)

    graphs[0, 0].set_title("Training Loss")
    graphs[0, 0].set(xlabel="batch #", ylabel="loss")
    graphs[0, 1].set(xlabel="batch #", ylabel="accuracy")
    graphs[0, 1].set_title("Training Accuracy")

    graphs[1, 0].set_title("Validation Loss (Covid Tweets)")
    graphs[1, 0].set(xlabel="batch #", ylabel="loss")
    graphs[1, 1].set(xlabel="batch #", ylabel="accuracy")
    graphs[1, 1].set_title("Validation Accuracy (Covid Tweets)")

    graphs[2, 0].set_title("Validation Loss (Non-Covid Tweets)")
    graphs[2, 0].set(xlabel="batch #", ylabel="loss")
    graphs[2, 1].set(xlabel="batch #", ylabel="accuracy")
    graphs[2, 1].set_title("Validation Accuracy (Non-Covid Tweets)")

    graphs[0, 0].plot(training_steps, training_losses)
    graphs[0, 1].plot(training_steps, training_accuracy, 'tab:orange')
    graphs[1, 0].plot(evaluation_steps, evaluation_losses, 'tab:green')
    graphs[1, 1].plot(evaluation_steps, evaluation_accuracy, 'tab:red')
    graphs[2, 0].plot(evaluation_steps, evaluation_losses_noncovid,
                      'tab:green')
    graphs[2, 1].plot(evaluation_steps, evaluation_accuracy_noncovid,
                      'tab:pink')

    graphs[1, 1].annotate("max: ({}, {:.3f})".format(xmax, ymax),
                          xy=(xmax, ymax),
                          xytext=(xmax - 5, ymax - 0.05))
    graphs[2, 1].annotate("max: ({}, {:.3f})".format(xmax2, ymax2),
                          xy=(xmax2, ymax2),
                          xytext=(xmax2 - 5, ymax2 - 0.04))

    plt.savefig('summary.png')
    #plt.show()


preprocess/preprocess.py
#!/usr/bin/python3
import sys
import pandas as pd
import concurrent.futures
import os

FILENAME = 'preprocess/training_set_raw.csv'
OUTPUT_POS = 'data/good_tweets.csv'
OUTPUT_NEG = 'data/bad_tweets.csv'


def isEnglish(s):
    try:
        s.encode(encoding='utf-8').decode('ascii')
    except UnicodeDecodeError:
        return False
    else:
        return True


def filter_non_english(start, end, data):
    print(f'processing {start} - {end-1}')
    droplist = []
    for i, row in data.iloc[start:end].iterrows():
        if not isEnglish(row['text']):
            droplist.append(i)
    return droplist


def main():
    # 1. load data
    data = pd.read_csv(
        FILENAME,
        encoding='latin_1',
        header=None,
        names=['label', 'tweet_id', 'date', 'idk', 'user', 'text'])
    # 2. drop unneeded columns
    data.drop(['tweet_id', 'date', 'idk', 'user'], axis=1, inplace=True)
    # 3. remove non-english tweets
    executor = concurrent.futures.ThreadPoolExecutor(8)
    step = 10000
    futures = [
        executor.submit(filter_non_english, s, min(s + step, len(data)), data)
        for s in range(0, len(data), step)
    ]
    done, pending = concurrent.futures.wait(futures)
    for f in done:
        data.drop(f.result(), inplace=True)

    # 4. split by positive/negative tweets
    pos = data[data['label'] == 4].drop('label', axis=1)
    neg = data[data['label'] == 0].drop('label', axis=1)
    # 4. output
    print(f'output {len(pos)} positive tweets and {len(neg)} negative tweets')
    if not os.path.exists('../data'):
        os.makedirs('../data')
    pos.to_csv(OUTPUT_POS, header=False, index=False)
    neg.to_csv(OUTPUT_NEG, header=False, index=False)


if __name__ == '__main__':
    main()


NN-algo/train.py
#! /usr/bin/env python
import sys

#SELECT WHICH MODEL YOU WISH TO RUN:
from cnn_lstm import CNN_LSTM  #OPTION 0
from lstm_cnn import LSTM_CNN  #OPTION 1
from cnn import CNN  #OPTION 2 (Model by: Danny Britz)
from lstm import LSTM  #OPTION 3
from graph_res import plot_results
MODEL_TO_RUN = 1

import tensorflow as tf
import numpy as np
import os
import time
import datetime
import batchgen
from tensorflow.contrib import learn
import statistics

from IPython import embed

# Parameters
# ==================================================

# Data loading params
dev_size = 0.1

# Model Hyperparameters
embedding_dim = 32  #128
max_seq_legth = 70
filter_sizes = [3, 4, 5]  #3
num_filters = 32
dropout_prob = 0.9  #0.5
l2_reg_lambda = 0.0
use_glove = False  #Do we use glove

# Training parameters
batch_size = 128
num_epochs = 10  #200
evaluate_every = 100  #100
checkpoint_every = 100000  #100
num_checkpoints = 0  #Checkpoints to store
MAX_TRAIN_SIZE = 20000

# Misc Parameters
allow_soft_placement = True
log_device_placement = False

# Parameters for plt
training_steps = []
training_losses = []
training_accuracy = []

evaluation_steps = []
evaluation_losses = []
evaluation_accuracy = []

evaluation_losses_noncovid = []
evaluation_accuracy_noncovid = []

# Data Preparation
# ==================================================

filename = "../tweets.csv"
goodfile = "../data/good_tweets.csv"
badfile = "../data/bad_tweets.csv"
goodvalidfile = "../data/good_validation.csv"
badvalidfile = "../data/bad_validation.csv"

# Load data
print("Loading data...")
x_text, y = batchgen.get_dataset(goodfile, badfile,
                                 MAX_TRAIN_SIZE)  #TODO: MAX LENGTH
# Build vocabulary
max_document_length = max([len(x.split(" ")) for x in x_text])
if (not use_glove):
    print("Not using GloVe")
    vocab_processor = learn.preprocessing.VocabularyProcessor(
        max_document_length)
    x = np.array(list(vocab_processor.fit_transform(x_text)))
else:
    print("Using GloVe")
    embedding_dim = 50
    filename = '../glove.6B.50d.txt'

    def loadGloVe(filename):
        vocab = []
        embd = []
        file = open(filename, 'r')
        for line in file.readlines():
            row = line.strip().split(' ')
            vocab.append(row[0])
            embd.append(row[1:])
        print('Loaded GloVe!')
        file.close()
        return vocab, embd

    vocab, embd = loadGloVe(filename)
    vocab_size = len(vocab)
    embedding_dim = len(embd[0])
    embedding = np.asarray(embd)

    W = tf.Variable(tf.constant(0.0, shape=[vocab_size, embedding_dim]),
                    trainable=False,
                    name="W")
    embedding_placeholder = tf.placeholder(tf.float32,
                                           [vocab_size, embedding_dim])
    embedding_init = W.assign(embedding_placeholder)

    session_conf = tf.ConfigProto(allow_soft_placement=True,
                                  log_device_placement=False)
    sess = tf.Session(config=session_conf)
    sess.run(embedding_init, feed_dict={embedding_placeholder: embedding})

    from tensorflow.contrib import learn
    #init vocab processor
    vocab_processor = learn.preprocessing.VocabularyProcessor(
        max_document_length)
    #fit the vocab from glove
    pretrain = vocab_processor.fit(vocab)
    #transform inputs
    x = np.array(list(vocab_processor.transform(x_text)))

    #init vocab processor
    vocab_processor = learn.preprocessing.VocabularyProcessor(
        max_document_length)
    #fit the vocab from glove
    pretrain = vocab_processor.fit(vocab)
    #transform inputs
    x = np.array(list(vocab_processor.transform(x_text)))

# Randomly shuffle data
np.random.seed(42)
shuffle_indices = np.random.permutation(np.arange(len(y)))
x_shuffled = x[shuffle_indices]
y_shuffled = y[shuffle_indices]

# Split train/test set
# TODO: This is very crude, should use cross-validation
dev_sample_index = -1 * int(dev_size * float(len(y)))
x_train, x_dev = x_shuffled[:dev_sample_index], x_shuffled[dev_sample_index:]
y_train, y_dev = y_shuffled[:dev_sample_index], y_shuffled[dev_sample_index:]

x_covid, y_covid = batchgen.get_dataset(goodvalidfile, badvalidfile,
                                        MAX_TRAIN_SIZE)
x_covid = np.array(list(vocab_processor.fit_transform(x_covid)))

print("Vocabulary Size: {:d}".format(len(vocab_processor.vocabulary_)))
print("Train/Dev split: {:d}/{:d}".format(len(y_train), len(y_dev)))

#embed()

# Training
# ==================================================

with tf.Graph().as_default():
    session_conf = tf.ConfigProto(allow_soft_placement=True,
                                  log_device_placement=False)
    sess = tf.Session(config=session_conf)
    with sess.as_default():
        #embed()
        if (MODEL_TO_RUN == 0):
            model = CNN_LSTM(x_train.shape[1], y_train.shape[1],
                             len(vocab_processor.vocabulary_), embedding_dim,
                             filter_sizes, num_filters, l2_reg_lambda)
        elif (MODEL_TO_RUN == 1):
            model = LSTM_CNN(x_train.shape[1], y_train.shape[1],
                             len(vocab_processor.vocabulary_), embedding_dim,
                             filter_sizes, num_filters, l2_reg_lambda)
        elif (MODEL_TO_RUN == 2):
            model = CNN(x_train.shape[1], y_train.shape[1],
                        len(vocab_processor.vocabulary_), embedding_dim,
                        filter_sizes, num_filters, l2_reg_lambda)
        elif (MODEL_TO_RUN == 3):
            model = LSTM(x_train.shape[1], y_train.shape[1],
                         len(vocab_processor.vocabulary_), embedding_dim)
        else:
            print(
                "PLEASE CHOOSE A VALID MODEL!\n0 = CNN_LSTM\n1 = LSTM_CNN\n2 = CNN\n3 = LSTM\n"
            )
            exit()

        # Define Training procedure
        global_step = tf.Variable(0, name="global_step", trainable=False)
        optimizer = tf.train.AdamOptimizer(1e-3)
        grads_and_vars = optimizer.compute_gradients(model.loss)
        train_op = optimizer.apply_gradients(grads_and_vars,
                                             global_step=global_step)

        # Keep track of gradient values and sparsity (optional)
        grad_summaries = []
        for g, v in grads_and_vars:
            if g is not None:
                grad_hist_summary = tf.summary.histogram(
                    "{}/grad/hist".format(v.name), g)
                sparsity_summary = tf.summary.scalar(
                    "{}/grad/sparsity".format(v.name), tf.nn.zero_fraction(g))
                grad_summaries.append(grad_hist_summary)
                grad_summaries.append(sparsity_summary)
        grad_summaries_merged = tf.summary.merge(grad_summaries)

        # Output directory for models and summaries
        timestamp = str(int(time.time()))
        out_dir = os.path.abspath(
            os.path.join(os.path.curdir, "runs", timestamp))
        print("Writing to {}\n".format(out_dir))

        # Summaries for loss and accuracy
        loss_summary = tf.summary.scalar("loss", model.loss)
        acc_summary = tf.summary.scalar("accuracy", model.accuracy)

        # Train Summaries
        train_summary_op = tf.summary.merge(
            [loss_summary, acc_summary, grad_summaries_merged])
        train_summary_dir = os.path.join(out_dir, "summaries", "train")
        train_summary_writer = tf.summary.FileWriter(train_summary_dir,
                                                     sess.graph)

        # Dev summaries
        dev_summary_op = tf.summary.merge([loss_summary, acc_summary])
        dev_summary_dir = os.path.join(out_dir, "summaries", "dev")
        dev_summary_writer = tf.summary.FileWriter(dev_summary_dir, sess.graph)

        # Checkpoint directory. Tensorflow assumes this directory already exists so we need to create it
        checkpoint_dir = os.path.abspath(os.path.join(out_dir, "checkpoints"))
        checkpoint_prefix = os.path.join(checkpoint_dir, "model")
        if not os.path.exists(checkpoint_dir):
            os.makedirs(checkpoint_dir)
        saver = tf.train.Saver(tf.global_variables(),
                               max_to_keep=num_checkpoints)

        # Write vocabulary
        vocab_processor.save(os.path.join(out_dir, "vocab"))

        # Initialize all variables
        sess.run(tf.global_variables_initializer())

        #TRAINING STEP
        def train_step(x_batch, y_batch, save=False):
            feed_dict = {
                model.input_x: x_batch,
                model.input_y: y_batch,
                model.dropout_keep_prob: dropout_prob
            }
            _, step, summaries, loss, accuracy = sess.run([
                train_op, global_step, train_summary_op, model.loss,
                model.accuracy
            ], feed_dict)
            time_str = datetime.datetime.now().isoformat()
            print("{}: step {}, loss {:g}, acc {:g}".format(
                time_str, step, loss, accuracy))

            # Training step
            if step % 5 == 0:
                training_steps.append(step)
                training_losses.append(loss)
                training_accuracy.append(accuracy)

            if save:
                train_summary_writer.add_summary(summaries, step)

        #EVALUATE MODEL
        def dev_step(x_batch,
                     y_batch,
                     writer=None,
                     save=False,
                     evaluate_step_general_tweets=False):
            feed_dict = {
                model.input_x: x_batch,
                model.input_y: y_batch,
                model.dropout_keep_prob: 0.5
            }
            step, summaries, loss, accuracy = sess.run(
                [global_step, dev_summary_op, model.loss, model.accuracy],
                feed_dict)
            time_str = datetime.datetime.now().isoformat()
            print("{}: step {}, loss {:g}, acc {:g}".format(
                time_str, step, loss, accuracy))

            if not evaluate_step_general_tweets:
                # Evaluation step using covid tweetes

                evaluation_steps.append(step)
                evaluation_losses.append(loss)
                evaluation_accuracy.append(accuracy)
            else:
                # Using non-covid tweets

                evaluation_losses_noncovid.append(loss)
                evaluation_accuracy_noncovid.append(accuracy)

            if save:
                if writer:
                    writer.add_summary(summaries, step)

            return loss, accuracy

        #CREATE THE BATCHES GENERATOR
        batches = batchgen.gen_batch(list(zip(x_train, y_train)), batch_size,
                                     num_epochs)

        # RECORD BEST RESULT
        highest_acc = -1
        record_loss = 0
        record_step = 0

        highest_acc_noncovid = -1
        record_loss_noncovid = 0
        record_step_noncovid = 0

        #TRAIN FOR EACH BATCH
        for batch in batches:
            x_batch, y_batch = zip(*batch)
            train_step(x_batch, y_batch)
            current_step = tf.train.global_step(sess, global_step)
            if current_step % evaluate_every == 0:
                print("\nEvaluation for general tweets:")
                step_loss, step_accuracy = dev_step(
                    x_dev,
                    y_dev,
                    writer=dev_summary_writer,
                    evaluate_step_general_tweets=True)

                if step_accuracy > highest_acc_noncovid:
                    highest_acc_noncovid = step_accuracy
                    record_loss_noncovid = step_loss
                    record_step_noncovid = current_step

                print("")
                print("\nEvaluation for the covid tweets:")
                step_loss, step_accuracy = dev_step(x_covid,
                                                    y_covid,
                                                    writer=dev_summary_writer)

                if step_accuracy > highest_acc:
                    highest_acc = step_accuracy
                    record_loss = step_loss
                    record_step = current_step

                print("")
            if current_step % checkpoint_every == 0:
                path = saver.save(sess,
                                  checkpoint_prefix,
                                  global_step=current_step)
                print("Saved model checkpoint to {}\n".format(path))

        dev_step(x_dev,
                 y_dev,
                 writer=dev_summary_writer,
                 evaluate_step_general_tweets=True)
        dev_step(x_covid,
                 y_covid,
                 writer=dev_summary_writer,
                 evaluate_step_general_tweets=False)

print("General tweets: step {}, loss {:g}, acc {:g}".format(
    record_step_noncovid, record_loss_noncovid, highest_acc_noncovid))
print("Covid-related tweets: step {}, loss {:g}, acc {:g}".format(
    record_step, record_loss, highest_acc))
print(f'min covid-loss {min(evaluation_losses)}')
print(f'min non-covid-loss {min(evaluation_losses_noncovid)}')

print(f'avg covid accuracy {statistics.mean(evaluation_accuracy)}')
print(
    f'avg non-covid accuracy {statistics.mean(evaluation_accuracy_noncovid)}')
plot_results(training_steps, training_losses, training_accuracy,
             evaluation_steps, evaluation_losses, evaluation_accuracy,
             evaluation_losses_noncovid, evaluation_accuracy_noncovid,
             record_step, highest_acc, record_step_noncovid,
             highest_acc_noncovid)
